keps/sig-node/3759-sidecar-containers/README.md

```diff
@@ -0,0 +1,1291 @@
+
+<!--
+**Note:** This checklist is iterative and should be reviewed and updated every time this enhancement is being considered for a milestone.
+-->
+
+[kubernetes.io]: https://kubernetes.io/
+[kubernetes/enhancements]: https://git.k8s.io/enhancements
+[kubernetes/kubernetes]: https://git.k8s.io/kubernetes
+[kubernetes/website]: https://git.k8s.io/website
+
+## Summary
+
+Sidecar containers are a new type of containers that start among the Init
+containers, run through the lifecycle of the Pod and don’t block pod
+termination. Kubelet makes a best effort to keep it alive and running while
+other containers are running.
+
+## Motivation
+
+The concept of sidecar containers has been around since the early days of
+Kubernetes. A clear example is [this Kubernetes blog post](https://kubernetes.io/blog/2015/06/the-distributed-system-toolkit-patterns/#example-1-sidecar-containers)
+from 2015 mentioning the sidecar pattern.
+
+Over the years the sidecar pattern has become more common in applications,
+gained popularity and the use cases are getting more diverse. The current
+Kubernetes primitives handle that well, but they are starting to fall short for
+several use cases and force weird work-arounds in the applications.
+
+The next sections expand on what the current problems are. But, to give more
+context, it is important to highlight that some companies are already using a
+fork of Kubernetes with this sidecar functionality added (not all
+implementations are the same, but more than one company has a fork for this).
+
+### Problems: jobs with sidecar containers
+
+Imagine you have a Job with two containers: one which does the main processing
+of the job and the other is just a sidecar facilitating it. This sidecar could
+be a service mesh, a metrics gathering statsd server, etc.
+
+When the main processing finishes, the pod won't terminate until the sidecar
+container finishes too. This is problematic for sidecar containers that run
+continuously.
+
+There is no simple way to handle this on Kubernetes today. There are
+work-arounds for this problem, most of them consist of some form of coupling
+between the containers to add some logic where a container that finishes
+communicates it so other containers can react. But it gets tricky when you have
+more than one sidecar container or want to auto-inject sidecars.
+
+The sidecar will also not be restarted when it was OOM killed, which may render
+the pod unusable if the sidecar provided secure communication to other
+containers. 
+
+### Problems: logging/metrics sidecar
+
+A logging sidecar should start before several other containers, to not lose logs
+from the startup of other applications. Let's call "main" container the app that
+will log and "logging" container the sidecar that will facilitate it.
+
+If the logging container starts after the main app, some logs can be lost.
```

Comment:
Logging is still an important scenario. We do not guarantee logs will be delivered, but making it significantly easier to implement in majority of cases. I guess there are gradations: from reliable logs collection to easy to implemet startup logs upload (that is 80% reliable) to no startup logs or hard to implement startup logs.

Would moving this scenario to "use cases" section instead of Motivation be better?

Reply:
I think I would just acknowledge that this is not a complete solution to the problem of sidecar reliability.

Reply:
I think this is probably c&p from the old KEP I wrote. I think this sentence is not so important, it was to make clearer the next example (the next sentence).

Reply:
it is a copy paste indeed =) I chenged wording a bit to remove the "lost" language that implies reliability improvements